### PR TITLE
fix #2318 Prevent FlatMapInner to request if fused in SYNC mode

### LINKED ISSUE #2318: NoSuchElementException in FlatMapMain.drainLoop under high load
We are using a combination of `flatMap`, `publishOn` and `Flux.fromStream` in our processing pipeline and see regular `NoSuchElementExceptions` that we couldn't really explain. The processing pipeline roughly looks like this:

```
hotSource
  .flatMap(data -> Flux.stream(deserializeDataToJavaStream(data))
  .publishOn(Schedulers.parallel())
  .subscribe()
```

The exception is not very frequent (once or twice per day), but it appears regularly (we've seen it for months now). See the example below for a full stacktrace.

To give some context: the application is under high load (millions of messages processed per second) and on some hosts cpu usage regularly reaches 100%. So a theoretical concurrency problem is likely to surface in this environment.

## Expected Behavior
No `NoSuchElementException` is thrown.

## Actual Behavior
`NoSuchElementException` is thrown because `iterator.next()` is called although the the iterator is already empty.

## Steps to Reproduce

With a small trick we were able to reproduce the problem in a unit test:
```java
  @Test
  public void reproduceNoSuchElementException
    final ExecutorService executor = Executors.newSingleThreadExecutor();
    Flux.<Long>create(sink ->
        sink.onRequest(requested ->
            executor.submit(() ->
                LongStream.range(0, requested).forEach(sink::next))))
        .flatMap(
            x -> Flux.fromStream(IntStream.range(0, 64).boxed()),
            8, 4)
        .publishOn(Schedulers.parallel())
        .take(Duration.ofSeconds(10))
        .blockLast();
```

If nothing else is change, this test runs fine. However, if I introduce a `Thread.yield()` in https://github.com/reactor/reactor-core/blob/v3.3.2.RELEASE/reactor-core/src/main/java/reactor/core/publisher/FluxFlatMap.java#L952 (between `queue=f` and `done=true`) then suddenly the test fails with the following exception:
```
java.util.NoSuchElementException
	at java.util.Spliterators$1Adapter.next(Spliterators.java:688)
	at reactor.core.publisher.FluxIterable$IterableSubscription.slowPath(FluxIterable.java:230)
	at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:201)
	at reactor.core.publisher.FluxFlatMap$FlatMapInner.request(FluxFlatMap.java:994)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drainLoop(FluxFlatMap.java:710)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.drain(FluxFlatMap.java:569)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.request(FluxFlatMap.java:339)
	at reactor.core.publisher.FluxContextStart$ContextStartSubscriber.request(FluxContextStart.java:132)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.runAsync(FluxPublishOn.java:405)
	at reactor.core.publisher.FluxPublishOn$PublishOnSubscriber.run(FluxPublishOn.java:484)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Except for some more intermediate reactor operators, this is exactly the stacktrace that we are also seeing in our production environment.

## Possible Solution

I don't have a solution. But at least an explanation. It is generally possible, that the `FluxFlatMap.FlatMapMain.onNext` and `FluxFlatMap.FlatMapMain.request` are called concurrently (the first because the upstream Publisher emits new elements, the second because downstream subscriber requests new elements). 

This means, that it's possible for thread-1 to execute `FluxFlatMap.FlatMapInner.onSubscribe`, while thread-2 executes `FluxFlatMap.FlatMapMain.drainLoop`.

This means, that it can happen that thread-2 sees a `inner` with `inner.queue != null` but with `inner.done == false`. Because `inner.queue` has already the new value assigned in `FlatMapInner.onSubscribe` because of operator fusion, but the `inner.done` hasn't been assigned yet, because thread-1 didn't execute this line yet. 

In this case, it can happen that during the `drainLoop` both methods `FluxIterable.IterableSubscription.poll` and `FluxIterable.IterableSubscription.request` are called (which shouldn't be possible, because the first should only be called if fusion is enabled while the second should only be called when fusion is disabled). And this can lead to the observed `NoSuchElementException`. 

## Your Environment

* Reactor version(s) used: 3.3.2.RELEASE
* JVM version (`java -version`):
```
openjdk version "11.0.7" 2020-04-14 LTS
OpenJDK Runtime Environment Zulu11.39+15-CA (build 11.0.7+10-LTS)
OpenJDK 64-Bit Server VM Zulu11.39+15-CA (build 11.0.7+10-LTS, mixed mode)
```
* OS and version (`uname -a`):
```
Linux filler 4.14.186-146.268.amzn2.x86_64 #1 SMP Tue Jul 14 18:16:52 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by simonbasle**

### CODE REVIEW

File: reactor-core/src/test/java/reactor/core/publisher/FluxFlatMapTest.java [hunk 1]
```diff
@@ -46,6 +59,81 @@
...
+				if ((requestedMode & Fuseable.SYNC) != 0) {
+					return Fuseable.SYNC;
+				}
+				return Fuseable.NONE;
+			}
+		};
+		inner.addAll(Arrays.asList(1, 2, 3, 4, 5));
+
+		FluxFlatMap.FlatMapMain<?, Integer> parent = new FluxFlatMap.FlatMapMain<>(
+				AssertSubscriber.create(0), v -> Mono.empty(), false, 8, Queues.small(), 4, Queues.small());
```
Comment (bsideup) [inline reactor-core/src/test/java/reactor/core/publisher/FluxFlatMapTest.java:127]: nit: WDYT about doing "one arg per line" here?

Comment (simonbasle) [reply reactor-core/src/test/java/reactor/core/publisher/FluxFlatMapTest.java]: done. once approved I'll squash the last 3 commits to merge with rebase

**Comment by simonbasle** [comment]
Since we'll release Californium soon (with important bugfixes in netty), I think this one could be embedded in `3.2.x`. wdyt @bsideup?

**Comment by simonbasle** [comment]
(rebased and retargeted towards 3.2.x)

**Pull request merged**